### PR TITLE
hwmv2: port STM32H7 SoCs and boards

### PR DESCRIPTION
**Pull request by erwango**

### CODE REVIEW

**Comment by erwango** [comment]
Doc update remains

File: boards/arduino/arduino_giga_r1/Kconfig.arduino_giga_r1 [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ARDUINO_GIGA_R1_STM32H747XX_M7
+	bool
```
Comment (erwango) [inline boards/arduino/arduino_giga_r1/Kconfig.arduino_giga_r1:5]: @tejlmand, @nordicjm On dual core target, I had to leave (/add) the bool to board name definition, as the default generated (I guess it is generated in other cases) Kconfig BOARD symbol is `CONFIG_BOARD_STM32H747I_DISCO`.
Is that the expected way to proceed ?

**Review by fabiobaltieri — APPROVED** [review]

**Review by ajarmouni-st — APPROVED** [review]

**Review by ajarmouni-st — COMMENTED** [review]
There are mentions of stm32h747i_disco_mx in:
- samples/boards/stm32/h7_dual_core/README.rst
- samples/modules/lvgl/demos/sample.yaml 
- samples/subsys/display/lvgl/sample.yaml
as well

**Review by ajarmouni-st — CHANGES_REQUESTED** [review]
There is a mention of stm32h747i_disco_m7 in boards/shields/st_b_lcd40_dsi1_mb1166/doc/index.rst as well :)

**Review by ajarmouni-st — APPROVED** [review]

**Comment by ajarmouni-st** [comment]
> There is a mention of stm32h747i_disco_m7 in boards/shields/st_b_lcd40_dsi1_mb1166/doc/index.rst as well :)

to be ignored

**Comment by nordicjm** [comment]
Needs rebase

File: soc/st/stm32/stm32h7/Kconfig [hunk 2]
```diff
@@ -4,15 +4,22 @@
...
+	select CPU_HAS_ICACHE if CPU_CORTEX_M7
+	select CPU_HAS_DCACHE if CPU_CORTEX_M7
+
+config STM32H7_DUAL_CORE
+	bool "Dual Core"
+	depends on SOC_SERIES_STM32H7X
+
+config STM32H7_BOOT_M4_AT_INIT
+	bool "Boot M4 core during M7 init independent of option byte BCM4."
+	default y
```
Comment (nordicjm) [inline soc/st/stm32/stm32h7/Kconfig:25]: `depends on SOC_SERIES_STM32H7X`?

File: boards/arduino/arduino_giga_r1/board.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+board:
+  name: arduino_giga_r1
+  vendor: arduino
+  socs:
+    - name: stm32h747xx
```
Comment (nordicjm) [inline boards/arduino/arduino_giga_r1/board.yml:5]: The python script is buggy, the dash should be aligned to the socs 's' character

File: boards/st/nucleo_h745zi_q/Kconfig.defconfig [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# STM32H745ZI Nucleo board configuration
+
+# Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M7 || BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M4
```
Comment (nordicjm) [inline boards/st/nucleo_h745zi_q/Kconfig.defconfig:6]: For these you can just do `if BOARD_NUCLEO_H745ZI_Q`

File: boards/arduino/arduino_giga_r1/board.yml (hunk 3 above)
Comment (erwango) [reply boards/arduino/arduino_giga_r1/board.yml:5]: So it has to be fixed in all stm32 based boards that have been merged till now. A clean up PR will be required. I propose at that point.

File: soc/st/stm32/stm32h7/Kconfig (hunk 2 above)
Comment (erwango) [reply soc/st/stm32/stm32h7/Kconfig:25]: Good catch

File: boards/st/nucleo_h745zi_q/Kconfig.defconfig (hunk 4 above)
Comment (erwango) [reply boards/st/nucleo_h745zi_q/Kconfig.defconfig:6]: indeed, txs

File: soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_STM32H745XX_M7 || SOC_STM32H745XX_M4
+
+config CPU_CORTEX_M7
+	default y if SOC_STM32H745XX_M7
+
+config CPU_CORTEX_M4
+	default y if SOC_STM32H745XX_M4
```
Comment (nordicjm) [inline soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx:12]: would like @tejlmand to comment on this, we are doing this in `Kconfig` for every other board as a select

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx (hunk 5 above)
Comment (erwango) [reply soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx:12]: This was indeed my initial version. I finally opted for this scheme as there files are there for SoC specific configuration

Comment (erwango) [reply soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx:12]: @tejlmand This is blocking stm32u5 and stm32wba migration

Comment (erwango) [reply soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx]: @nordicjm PTAL

Comment (tejlmand) [reply soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx]: > @tejlmand This is blocking stm32u5 and stm32wba migration

I see code has changed to `select`, which would also be my preference.

The `CPU_CORTEX_Mx` settings are promptless and thus not suffering the stuck symbol syndrome.
Also it's tightly coupled to the SoC selection, so make sense to use select.

Comment (erwango) [reply soc/st/stm32/stm32h7/Kconfig.defconfig.stm32h745xx]: Ok, we'll keep it that way then. Thanks

File: boards/arduino/arduino_giga_r1/board.yml (hunk 3 above)
Comment (tejlmand) [reply boards/arduino/arduino_giga_r1/board.yml:5]: > A clean up PR will be required. I propose at that point.

I'm fine with that, then we have a common cleanup PR for all STM.

File: boards/arduino/arduino_giga_r1/Kconfig.arduino_giga_r1 [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+	select SOC_STM32H747XX_M7
+
+config BOARD_ARDUINO_GIGA_R1_STM32H747XX_M4
+	bool
+	select SOC_STM32H747XX_M4
```
Comment (tejlmand) [inline boards/arduino/arduino_giga_r1/Kconfig.arduino_giga_r1:10]: > Is that the expected way to proceed ?

No, please do like this:
```suggestion
config BOARD_ARDUINO_GIGA_R1
	select SOC_STM32H747XX_M7 if BOARD_ARDUINO_GIGA_R1_STM32H747XX_M7
	select SOC_STM32H747XX_M4 if BOARD_ARDUINO_GIGA_R1_STM32H747XX_M4
```

this way it's ensured that Kconfig will fail with undefined type for symbol if board name is mismatching / renamed, as well as CI will fail if the `select <SOC>` or `if BOARD_<full_identifier>` are wrong / renamed.

File: boards/arduino/arduino_opta/Kconfig.arduino_opta [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2023 Felipe Neves
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ARDUINO_OPTA_STM32H747XX_M4
+	bool
```
Comment (tejlmand) [inline boards/arduino/arduino_opta/Kconfig.arduino_opta:5]:
```suggestion
```

File: boards/arduino/arduino_opta/Kconfig.arduino_opta [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2023 Felipe Neves
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ARDUINO_OPTA_STM32H747XX_M4
```
Comment (tejlmand) [inline boards/arduino/arduino_opta/Kconfig.arduino_opta:4]:
```suggestion
config BOARD_ARDUINO_OPTA
```

Please use the pure board name identifier, as that will ensure the correct soc is still selected if variants are later added to the board.

File: boards/arduino/arduino_portenta_h7/Kconfig.arduino_portenta_h7 [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) 2022 Benjamin Björnsson <benjamin.bjornsson@gmail.com>.
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ARDUINO_PORTENTA_H7_STM32H747XX_M7
+	bool
+	select SOC_STM32H747XX_M7
+
+config BOARD_ARDUINO_PORTENTA_H7_STM32H747XX_M4
+	bool
+	select SOC_STM32H747XX_M4
```
Comment (tejlmand) [inline boards/arduino/arduino_portenta_h7/Kconfig.arduino_portenta_h7:12]:
```suggestion
config BOARD_ARDUINO_PORTENTA_H7
	select SOC_STM32H747XX_M7 if BOARD_ARDUINO_PORTENTA_H7_STM32H747XX_M7
	select SOC_STM32H747XX_M4 if BOARD_ARDUINO_PORTENTA_H7_STM32H747XX_M4
```

File: boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg
Comment (tejlmand) [inline boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg:1]: why is this file being removed and added again.

I do notice that the new file is only 58kB, whereas the previous file was 177 kB, but due to the nature of git and the way fetch works, then any user doing a git fetch / clone on Zephyr repo will now get both the new 55 kB + old 177 kB file, meaning no performance gain.

If just keeping old 177 kB file, then users only need to fetch less data.

So if reason is size, then just stick with old file.
If there are other reasons, then of course update the picture.

File: boards/st/nucleo_h723zg/doc/img/nucleo_h723zg.jpg
Comment (tejlmand) [inline boards/st/nucleo_h723zg/doc/img/nucleo_h723zg.jpg:1]: same comment on file size.

File: boards/st/nucleo_h745zi_q/Kconfig.nucleo_h745zi_q [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+# Copyright (c) 2020 Alexander Kozhinov <AlexanderKozhinov@yandex.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M7
+	bool
+	select SOC_STM32H745XX_M7
+
+config BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M4
+	bool
+	select SOC_STM32H745XX_M4
```
Comment (tejlmand) [inline boards/st/nucleo_h745zi_q/Kconfig.nucleo_h745zi_q:12]:
```suggestion
config BOARD_NUCLEO_H745ZI_Q
	select SOC_STM32H745XX_M7 if BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M7
	select SOC_STM32H745XX_M4 if BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M4
```

**Review by tejlmand — DISMISSED** [review]

File: boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg
Comment (erwango) [reply boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg:1]: Reason is CI complaining for large file.

Comment (nordicjm) [reply boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg:1]: CI for large images can be ignored, re-use the existing image as to not further increase repo size

File: boards/st/nucleo_h723zg/doc/img/nucleo_h723zg.jpg
Comment (nordicjm) [reply boards/st/nucleo_h723zg/doc/img/nucleo_h723zg.jpg:1]: And here

File: boards/st/nucleo_h745zi_q/board.cmake [hunk 11]
```diff
@@ -1,9 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 board_runner_args(jlink "--device=STM32H745ZI" "--speed=4000")
-if(CONFIG_BOARD_NUCLEO_H745ZI_Q_M7)
+if(BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M7)
 board_runner_args(openocd --target-handle=_CHIPNAME.cpu0)
-elseif(CONFIG_BOARD_NUCLEO_H745ZI_Q_M4)
+elseif(BOARD_NUCLEO_H745ZI_Q_STM32H745XX_M4)
```
Comment (nordicjm) [inline boards/st/nucleo_h745zi_q/board.cmake:6]: missing CONFIG_ from these

File: boards/st/nucleo_h7a3zi_q/doc/img/nucleo_h7a3zi_q.jpg
Comment (nordicjm) [inline boards/st/nucleo_h7a3zi_q/doc/img/nucleo_h7a3zi_q.jpg:1]: And here

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg
Comment (erwango) [reply boards/arduino/arduino_portenta_h7/doc/img/arduino_portenta_h7.jpeg:1]: So please fix CI then.

File: boards/st/nucleo_h745zi_q/board.cmake (hunk 11 above)
Comment (erwango) [reply boards/st/nucleo_h745zi_q/board.cmake:6]: Good catch. Txs

**Comment by nordicjm** [comment]
Rebase needed and fixes for doc build to pass

**Comment by erwango** [comment]
@nordicjm rebased PTAL

**Review by nordicjm — APPROVED** [review]

**Review by nordicjm — CHANGES_REQUESTED** [review]
Doc updates needed:
```
/runner/_work/zephyr/zephyr/doc/_build/src/boards/fanke/fk7b0m1_vbt6/doc/index.rst: WARNING: document isn't included in any toctree
```
Essentially copy the index.rst from e.g. boards/digilent and place in the vendor folders with updated name/title

**Review by nordicjm — APPROVED** [review]

**Comment by erwango** [comment]
@nordicjm This should be fine now. Are you able to merge ?

**Pull request merged**